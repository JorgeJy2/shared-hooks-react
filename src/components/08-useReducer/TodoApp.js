import React, { useCallback, useEffect, useReducer } from 'react';
import { Todos } from './Todos';
import { todoReducer } from './reducer/todoReducer';

import './styles.css';
import { AddTodo } from './AddTodo';

// const initialStateTodos = [{
//     id: new Date().getTime(),
//     describe: 'Buy milk',
//     done: false
// }];

const init = () => JSON.parse(localStorage.getItem('todos')) || [];

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => localStorage.setItem('todos', JSON.stringify(todos)), [todos])

    const handlerDelete = (id) =>
        dispatch({
            type: 'delete',
            payload: {
                id
            }
        });

    const handlerToggle = (id) =>
        dispatch({
            type: 'toggle',
            payload: {
                id
            }
        });

    const handlerAdd = useCallback(
        (todo) =>
        dispatch({
            type: 'add',
            payload: todo
        }), [dispatch]
    );


    return (
        <>
            <h1>Todo app <small>{todos.length}</small> </h1>
            <hr />

            <Todos todos={todos} handlerDelete={handlerDelete} handlerToggle={handlerToggle} />
            <AddTodo handlerAdd={handlerAdd} />
        </>
    )
}

