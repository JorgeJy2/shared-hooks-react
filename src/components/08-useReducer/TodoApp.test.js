import { mount, shallow } from "enzyme";
import { act } from "react-dom/test-utils";
import { TodoApp } from "./TodoApp";

const todos = [{
    id: 1,
    describe: 'Learn react',
    done: false
}, {
    id: 2,
    describe: 'Learn python',
    done: false
}];

describe('todo app test', () => {
    Storage.prototype.setItem = jest.fn(() => { });
    const wrapper = shallow(<TodoApp />);

    test('should snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should add', () => {
        const wrapper = mount(<TodoApp />);

        act(() => {
            wrapper.find('AddTodo').prop('handlerAdd')(todos[0]);
            wrapper.find('AddTodo').prop('handlerAdd')(todos[1]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('Todo app 2');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });


    test('should delete', () => {
        const wrapper = mount(<TodoApp />);

        act(() => {
            wrapper.find('AddTodo').prop('handlerAdd')(todos[0]);
        });

        expect(wrapper.find('h1').text().trim()).toBe('Todo app 1');
        act(() => {
            wrapper.find('Todos').prop('handlerDelete')(todos[0].id);
        });

        expect(wrapper.find('h1').text().trim()).toBe('Todo app 0');
    });

});