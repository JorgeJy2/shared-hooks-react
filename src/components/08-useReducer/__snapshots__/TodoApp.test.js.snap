// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo app test should snapshot 1`] = `
<Fragment>
  <h1>
    Todo app 
    <small>
      0
    </small>
     
  </h1>
  <hr />
  <Todos
    handlerDelete={[Function]}
    handlerToggle={[Function]}
    todos={Array []}
  />
  <AddTodo
    handlerAdd={[Function]}
  />
</Fragment>
`;
